File: frontend/src/constants/orderStatus.ts
import { OrderStatus } from '@/types/order';
import { Tone } from '@/types/ui';

export const ORDER_STATUS_LABEL: Record<OrderStatus, string> = {
  RECEIVED: '주문접수',
  OUTBOUND_WAITING: '출고대기',
  PICKING: '피킹중',
  PICKED: '피킹완료',
  PACKED: '포장완료',
  LOADED: '상차완료',
  DISPATCHED: '배차완료',
  IN_DELIVERY: '배송중',
  DELIVERED: '배송완료',
  FAILED: '배송실패',
};

export const ORDER_STATUS_TONE: Record<OrderStatus, Tone> = {
  RECEIVED: 'gray',
  OUTBOUND_WAITING: 'gray',
  PICKING: 'yellow',
  PICKED: 'yellow',
  PACKED: 'blue',
  LOADED: 'blue',
  DISPATCHED: 'blue',
  IN_DELIVERY: 'green',
  DELIVERED: 'green',
  FAILED: 'red',
};

File: backend/src/main/java/com/logistics/domain/order/domain/OrderStatus.java
package com.logistics.domain.order.domain;

/**
 * 주문접수 → 출고대기 → 피킹중 → 피킹완료 → 포장완료 → 상차완료 → 배차완료 → 배송중 → 배송완료 | 배송실패
 * 배차 취소 시에만 배차완료에서 상차완료로 되돌릴 수 있다.
 */
public enum OrderStatus {
    RECEIVED,
    OUTBOUND_WAITING,
    PICKING,
    PICKED,
    PACKED,
    LOADED,
    DISPATCHED,
    IN_DELIVERY,
    DELIVERED,
    FAILED;

    public boolean canMoveTo(OrderStatus to) {
        return switch (this) {
            case RECEIVED -> to == OUTBOUND_WAITING;
            case OUTBOUND_WAITING -> to == PICKING;
            case PICKING -> to == PICKED;
            case PICKED -> to == PACKED;
            case PACKED -> to == LOADED;
            case LOADED -> to == DISPATCHED;
            case DISPATCHED -> to == IN_DELIVERY || to == LOADED;
            case IN_DELIVERY -> to == DELIVERED || to == FAILED;
            case DELIVERED, FAILED -> false;
        };
    }
}
